### DISCUSSION TITLE
Able to blink or fade LED channels?

### DISCUSSION THREAD

**Discussion by mjpcomp** [tjko/brickpico#17] [General]
If we're building something that would look better with blinking lights (e.g. plane, some vehicles), is being able to either set a channel to a certain blink rate (1/sec, 2/sec, 1/min, etc...) or a fade rate (start on, fade down to 20% over 3 min, reverse so back to 100% after 3 min, or something like start off, fade up to 100% over 1 min and then turn off, wait 1 min, start again) feasible?

Not sure how complicated it would be, but sure would make for some amazing displays. I mean, could probably also do it for other devices other than LEGO, though the 200mA/channel would be a limiting factor for some other LED situations.

**Comment by tjko** [comment]
It's been on my "to do" list to add support for different types of "patterns" that could be selected for each channel separately.

One option would be to add some 'basic' modes linke "blink", "fade", etc that would take some parameters each.

- BLINK,brightness (%),on_duration (ms),off_duration (ms)
- FADER,start_brightness (%),end_brightness (%),fade_duration (ms),pause (ms),reverse (Y/N)
- ...

I was initially thinking allowing "programmable" patterns, but that would be quite a bit of work... but that might be something to add later after adding some basic modes first.

**Reply by st2000** [reply]
I was looking at this. Decided upon changing the on/off bool to something along the lines of on/off/pulse. Eventually wanting to get to [on,off,pulse[0,1,2,3]]. (i.e. Where pulse0 might be configured for period=1 sec, duty-cycle=10% & attack-decay=0 sec. And pulse1 might be configured for period=2 sec, duty-cycle=50% & attack-decay=0.5 sec) But this bool type found in structure brickpio_config is passed extensively between functions such that many variables (local_echo, spi_active, serial_active, mqtt_tls, mqtt_allow_scpi, telnet_active, telnet_auth & telnet_raw_mode) all need to be changed from bool to something else (i.e. char). That just doesn't sound right.

Regardless, I'm going to continue down this rabbit hole and see if I can get different styles of blinking working.  Then deal with the fall out, if any, later.

**Comment by st2000** [comment]
Experimenting with some blinking and fading on a modified version of brickpico. Many features are still hard coded. Not really ready for sharing yet.

https://github.com/user-attachments/assets/5c2a604f-ac21-4d4b-a915-a84969e131eb

**Reply by tjko** [reply]
Looks cool.   I had started work the initial "effects engine" work, let me see if I can get that completed and merged.  My plan is to use the second core to run the lighting "effects" and have simple plugins that can be easily added for new/different lighting effects...

**Reply by st2000** [reply]
I was planning on a merge request for these changes. But will stop if you already have something done. Duplicated efforts is just a wast of time.

**Comment by tjko** [comment]
Initial support for "effects" is now merged, there is three basic/simple effects included: fade (in/out), blink, and pulse: 52ba9d19f5a108cc0ecfcae3f335326d648d21b1

Lighting effects are "run" by second core so they should run rather consistently regardless how busy the first core is with networking, etc...  And it should be easy to add more effects.

(pre-built firmware can be found here: https://github.com/tjko/brickpico/actions/runs/12612639715)

**Reply by st2000** [reply]
I will have to try this! Thanks for creating this software & hardware.

**Comment by tjko** [comment]
Light output from LED lights is clearly not proportional to the PWM signal duty cycle, so I added initial support for Gamma correction: 47579236a20e7636d9028155fad3e2d422e54cab

Now the LED light brightness should more closely match the set output level (percentage), this would seem to make 'effects' look slightly better.

(pre-build "beta" firmware can be found here: https://github.com/tjko/brickpico/actions/runs/12703769226)

**Comment by st2000** [comment]
I pulled and built just moments ago. All your recent changes appear to be working as expected. Nice work!
